numbat/src/command.rs

```diff
@@ -0,0 +1,436 @@
+                }));
+            };
+
+            if words.next().is_some() {
+                return Some(Err(ParseError {
+                    kind: ParseErrorKind::InvalidCommand(err_msg),
+                    span: span_through_end(&word_boundaries, 1, code_source_id),
+                }));
+            }
+
+            Command::Info { item }
+        }
+        "list" | "ls" => {
+            let items = match words.next() {
+                None => None,
+                Some("functions") => Some(ListItems::Functions),
+                Some("dimensions") => Some(ListItems::Dimensions),
+                Some("variables") => Some(ListItems::Variables),
+                Some("units") => Some(ListItems::Units),
+                _ => {
+                    return Some(Err(ParseError {
+                        kind: ParseErrorKind::InvalidCommand(
+                            "if provided, the argument to `list` or `ls` must be \
+                             one of: functions, dimensions, variables, units",
+                        ),
+                        span: span_from_boundary(word_boundaries[1], code_source_id),
+                    }));
+                }
+            };
+            if words.next().is_some() {
+                let start = word_boundaries[2].0;
+                let end = word_boundaries.last().unwrap().1;
+                return Some(Err(ParseError {
+                    kind: ParseErrorKind::InvalidCommand("`list` takes at most one argument"),
+                    span: span_from_boundary((start, end), code_source_id),
+                }));
+            }
+
+            Command::List { items }
+        }
+        "save" => {
+            let err_msg = "`save` requires exactly one argument, the destination";
+            let Some(dst) = words.next() else {
+                return Some(Err(ParseError {
+                    kind: ParseErrorKind::InvalidCommand(err_msg),
+                    span: span_from_boundary(word_boundaries[0], code_source_id),
+                }));
+            };
+
+            if words.next().is_some() {
+                let start = word_boundaries[2].0;
+                let end = word_boundaries.last().unwrap().1;
+                return Some(Err(ParseError {
+                    kind: ParseErrorKind::InvalidCommand(err_msg),
+                    span: span_from_boundary((start, end), code_source_id),
+                }));
+            }
+
+            Command::Save { dst }
+        }
+
+        _ => return None,
+    };
+
+    Some(Ok(command))
+}
+
+/// Attempt to parse the input as a command, such as "help", "list <args>", "quit", etc
+///
+/// Returns:
+/// - `None`, if the input does not begin with a command keyword
+/// - `Some(Ok(Command))`, if the input is a valid command
+/// - `Some(Err(_))`, if the input starts with a valid command but has the wrong number
+///   or kind of arguments, e.g. `list foobar`
+pub fn parse_command<'a>(
+    input: &'a str,
+    resolver: &mut Resolver,
```

Comment:
Instead of passing in a `Resolver` reference, could we also pass in a code_source_id?

Reply:
Yes, definitely. That is much cleaner.

I realize however that there is a bug in this PR, which is that for non-command inputs we end up running `resolver.add_code_source` twice, which results in line numbers being off by a factor of ~2 for erroneous non-command inputs. (The precise factor depends on how many commands we've run, since those only trigger `resolve.add_code_source` once.) I'm running into a bit of a chicken and egg problem: to report errors in commands requires a code source up front, but we don't know if we're even parsing a command until we parse at all. An obvious solution is to double-parse — check if we have a command at all (probably quite cheaply, using `split_whitespace` to just grab the first word) and then if we do, fall back to a full parse with a `code_source_id`. This way we only call `resolve.add_code_source` at most once per input. Does this sound good?

Reply:
> I realize however that there is a bug in this PR, which is that for non-command inputs we end up running `resolver.add_code_source` twice, which results in line numbers being off by a factor of ~2 for erroneous non-command inputs. (The precise factor depends on how many commands we've run, since those only trigger `resolve.add_code_source` once.)

Minor: it's not really the line number that is wrong, but the "code source ID", right? The number in `<input:…>` angle brackets.

> I'm running into a bit of a chicken and egg problem: to report errors in commands requires a code source up front, but we don't know if we're even parsing a command until we parse at all. An obvious solution is to double-parse — check if we have a command at all (probably quite cheaply, using `split_whitespace` to just grab the first word) and then if we do, fall back to a full parse with a `code_source_id`. This way we only call `resolve.add_code_source` at most once per input. Does this sound good?

Hm. That would be a good reason to keep your old solution maybe? Or could you pass in a `get_code_source_id` callable that would only be executed in the error case?

Reply:
> Minor: it's not really the line number that is wrong, but the "code source ID", right? The number in <input:…> angle brackets.

Yes, that was the issue.

> Hm. That would be a good reason to keep your old solution maybe? Or could you pass in a get_code_source_id callable that would only be executed in the error case?

My old solution never actually worked. What I did was split the parser into a struct `SourcelessCommandParser` containing just the input, and then the `CommandParser` wraps that and a `code_source_id`. `SourcelessCommandParser` has a fallible initializer that only succeeds if we are in fact looking at a command line. Only if it succeeds do we go ahead and generate the `code_source_id`, and then we have everything we need to parse the command. The key is that you can't construct a `CommandParser` in the first place unless it's been proven that we have a valid command line (meaning it starts with a command word, not that all the arguments are ok), so we never unnecessarily generate a `code_source_id`.